README.md
# Flask DynamoDB Sessions

[![](https://img.shields.io/pypi/v/flask-dynamodb-sessions.svg)](https://pypi.org/project/flask-dynamodb-sessions/) [![Updates](https://pyup.io/repos/github/ibejohn818/flask-dynamodb-sessions/shield.svg)](https://pyup.io/repos/github/ibejohn818/flask-dynamodb-sessions/)

Server-side sessions in Flask using AWS DynamoDB as the backend data store.

DynamoDB is AWS's SaaS NoSQL solution which makes it perfect for use as a session store. 
Being a SaaS service we no longer have to manage servers/storage/etc and take advantage of some notable features such as:

- Auto-scaling
- Automatic Lifecycle ( Garbage collection )
- Encryption at rest
- etc...

Sessions are pickled and base64 encoded to be stored in DynamoDB as strings. As a result you may save
objects to your sessions as long as the object supports the pickle interface.

DynamoDB supports a maximum object size of 400 KB. Minus the UUID4 session id, modified date/time string and ttl timestamp
you have approximately 398 KB available for your session.

## Installation

```shell
# w/ pip
pip install flask-dynamodb-sessions
#  w/ easy_install
easy_install flask-dynamodb-sessions
```

### Usage example
```python
from flask import (Flask, session)
from flask_dynamodb_sessions import Session

app = Flask(__name__)

# Set flask to use the dynamo session interface
Session(app)

@app.route('/', methods=['GET'])
def index_get():
	# use sessions just as you normally would
	session['user'] = {'username': 'jhardy'}
	
	user = session.get('user')
	
	session_id = session.sid
```
*View examples directory for more*

### Configuration Options
Below are additional `SESSION_*` configuration options specific to DynamoDB sessions.

    SESSION_DYNAMODB_TABLE (string): The DynamoDB table to save to. Default: flask_sessions
    SESSION_DYNAMODB_ENDPOINT (string): Override the boto3 endpoint, good for local development and using dynamodb-local. Default: None
    SESSION_DYNAMODB_TTL_SECONDS (int): Number of seconds to add to the TTL column. Default: 86400 * 14 (14 Days)

The existing `SESSION_*` config parameters still apply (IE: cookie settings). SESSION_REFRESH_EACH_REQUEST 
is the only setting that is negated and each request will refesh the cookie (Might be modified in a future release).

### Table Structure
The table structure is fairly simple.
```
{
    id: string HASH,
    modified: string DATETIME UTC
    ttl: number UTC TIME + SESSION_DYANMODB_TTL_SECONDS
    data: string JSON ENCODED SESSION

}
```

Create the table VIA `aws` cli.

```
aws dynamodb create-table --key-schema "AttributeName=id,KeyType=HASH" \
--attribute-definitions "AttributeName=id,AttributeType=S" \
--provisioned-throughput "ReadCapacityUnits=5,WriteCapacityUnits=5" \
--table-name flask_sessions
```

The `ttl` column is present to take advantage of DynamoDB's `Lifecycle` feature where dynamo will delete all rows with a ttl in the past.

Enable time-to-live (garbage collection)

```
aws dynamodb update-time-to-live --time-to-live-specification 'Enabled=true,AttributeName=ttl' --table-name flask_sessions
```



## TODO
- Test coverage
- More laxed cookie refresh


HISTORY.rst
=======
History
=======

0.1.0 (2019-01-04)
------------------

* First release on PyPI.


tests/test_flask_dynamodb_sessions.py
import pytest
from pytest_mock import mocker
import flask
from flask_dynamodb_sessions import Session


def test_session_boto_settings(mocker):

    client_mock = mocker.patch('flask_dynamodb_sessions.boto3.client')

    app = flask.Flask(__name__)

    app.config.update(
        SESSION_DYNAMODB_REGION='bogus-region',
        SESSION_DYNAMODB_ENDPOINT='http://bogus:1234'
    )


flask_dynamodb_sessions/__init__.py
# -*- coding: utf-8 -*-

"""
Flask DynamoDB Sessions

This module will use AWS DynamoDB as your session store in Flask.

"""
import sys
import time
from datetime import datetime
from uuid import uuid4
from flask.sessions import SessionInterface
from flask.sessions import SessionMixin
from werkzeug.datastructures import CallbackDict
import json
import boto3

PV3 = sys.version_info[0] == 3

import pickle
import codecs

__author__ = """John Hardy"""
__email__ = 'john@johnchardy.com'
__version__ = '0.1.4'


class Session(object):

    def __init__(self, app=None, **kw):
        self.app = app
        self.permanent = kw.get('permanent', True)

        if self.app is not None:
            self.init(self.app)

    def init(self, app):

        conf = app.config.copy()

        conf.setdefault("SESSION_DYNAMODB_ENDPOINT", None)
        conf.setdefault("SESSION_DYNAMODB_REGION", None)
        conf.setdefault("SESSION_DYNAMODB_TABLE", 'flask_sessions')
        conf.setdefault("SESSION_DYNAMODB_TTL_SECONDS", (86400 * 14))

        kw = {
            'table': conf['SESSION_DYNAMODB_TABLE'],
            'endpoint': conf['SESSION_DYNAMODB_ENDPOINT'],
            'region': conf['SESSION_DYNAMODB_REGION'],
            'ttl': conf['SESSION_DYNAMODB_TTL_SECONDS'],
            'permanent': self.permanent,
        }

        interface = DynamodbSessionInterface(**kw)

        app.session_interface = interface

class DynamodbSession(CallbackDict, SessionMixin):

    def __init__(self, initial=None, sid=None, permanent=None):
        def on_update(self):
            self.modified = True
        CallbackDict.__init__(self, initial, on_update)
        self.sid = sid
        if permanent:
            self.permanent = permanent
        self.modified = False


class DynamodbSessionInterface(SessionInterface):
    """
    """
    _boto_client = None

    def __init__(self, **kw):
        self.table = kw.get('table', 'flask_sessions')
        self.permanent = kw.get('permanent', True)
        self.endpoint = kw.get('endpoint', None)
        self.region = kw.get('region', None)
        self.ttl = kw.get('ttl', None)

    def open_session(self, app, req):
        """
        """
        id = req.cookies.get(app.session_cookie_name)

        if id is None:
            id = str(uuid4())
            return DynamodbSession(sid=id, permanent=self.permanent)

        data = self.dynamo_get(id)

        if data is not None:
            data = self.hydrate_session(data)

        return DynamodbSession(data, sid=id, permanent=self.permanent)

    def save_session(self, app, session, res):
        """
        """
        domain = self.get_cookie_domain(app)
        path = self.get_cookie_path(app)
        if not session:
            if session.modified:
                self.delete_session(session.sid)
                res.delete_cookie(app.session_cookie_name,
                                       domain=domain, path=path)
            return

        # if not self.should_set_cookie(app, session):
        #    return

        httponly = self.get_cookie_httponly(app)
        secure = self.get_cookie_secure(app)
        expires = self.get_expiration_time(app, session)
        session_id = session.sid

        self.dynamo_save(session_id, dict(session))

        res.set_cookie(app.session_cookie_name, session_id,
                            expires=expires, httponly=httponly,
                            domain=domain, path=path, secure=secure)



    def pickle_session(self, session):
        """Pickle the session object and base64 encode it
            for storage as a dynamo string
        """
        pickled = pickle.dumps(session)

        canned = codecs.encode(pickled, 'base64').decode()

        return canned

    def hydrate_session(self, session_data):
        """Base64 decode string back to bytes and unpickle
        """
        uncanned = codecs.decode(session_data.encode(), 'base64')

        pickled = pickle.loads(uncanned)

        return pickled

    def dynamo_get(self, session_id):
        """
        """
        try:
            res = self.boto_client().get_item(TableName=self.table,
                        Key={'id':{'S': session_id}})
            if res.get('Item').get('data'):
                data = res.get('Item').get('data')
                return data.get('S', '{}')
        except Exception as e:
            print("DYNAMO SESSION GET ITEM ERR: ", str(e))

        return None

    def dynamo_save(self, session_id, session):
        try:
            print(session)
            fields = {
                'data': {'S': self.pickle_session(session)},
                'modified': {'S': str(datetime.utcnow())},
                'ttl': {'N': str(int(datetime.utcnow().timestamp() + self.ttl))}
            }

            attr_names = {}
            attr_vals = {}
            ud_exp = []
            for k, v in fields.items():
                attr = "#attr_{}".format(k)
                token = ":{}".format(k)
                ud_exp.append("{} = {}".format(attr, token))
                attr_vals[token] = v
                attr_names[attr] = k

            self.boto_client().update_item(TableName=self.table,
                        Key={'id':{'S':session_id}},
                        ExpressionAttributeNames=attr_names,
                        ExpressionAttributeValues=attr_vals,
                        UpdateExpression="SET {}".format(", ".join(ud_exp)),
                        ReturnValues='NONE')
        except Exception as e:
            print("DYNAMO SESSION SAVE ERR: ", str(e))

    def delete_session(self, session_id):
        try:
            self.boto_client().delete_item(TableName=self.table,
                        Key={'id':{'S':session_id}})
        except Exception as e:
            print("DYNAMO SESSION DELETE ERR: ", str(e))

    def boto_client(self):
        """
        """
        if self._boto_client is None:
            kw = {}

            if self.endpoint is not None:
                kw['endpoint_url'] = self.endpoint
            if self.region is not None:
                kw['region_name'] = self.region

            self._boto_client = boto3.client('dynamodb', **kw)

        return self._boto_client
